File: core/src/de/iss/mv2/security/AESKeyGenerator.java
package de.iss.mv2.security;

import java.security.SecureRandom;

/**
 * A class for generating random AES-keys.
 * @author Marcel Singer
 *
 */
public class AESKeyGenerator implements SymmetricKeyGenerator{

	/**
	 * A field to hold a fixed key.
	 */
	private byte[] fixedKey;
	/**
	 * A field to hold a fixed iv.
	 */
	private byte[] fixedIV;
	
	/**
	 * Creates a new instance of {@link AESKeyGenerator}.
	 */
	public AESKeyGenerator() {
	}

	
	
	@Override
	public byte[] getRandomKey(int length) {
		if(fixedKey != null && fixedKey.length == length) return fixedKey;
		SecureRandom sr = new SecureRandom();
		byte[] result = new byte[length];
		sr.nextBytes(result);
		return result;
	}

	@Override
	public byte[] getRandomIV() {
		if(fixedIV != null) return fixedIV;
		return getRandomKey(16);
	}

	@Override
	public void setFixedKey(byte[] key) {
		this.fixedKey = key;
	}

	@Override
	public void setFixedIV(byte[] iv) {
		this.fixedIV = iv;
	}

	@Override
	public boolean hasFixedKeyAndIV() {
		return (fixedKey != null && fixedIV != null);
	}



	@Override
	public byte[] getRandomKey() {
		if(fixedKey != null) return fixedKey;
		return getRandomKey(16);
	}



	@Override
	public SymmetricKeyGenerator doClone() {
		AESKeyGenerator cl = new AESKeyGenerator();
		return cl;
	}

}

File: WebClient/MV2CorePortable/Security/AESKeyGenerator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace ISS.MV2.Security {
    public class AESKeyGenerator : ISymmetricKeyGenerator {

        private byte[] fixedKey = null;
        private byte[] fixedIV = null;

        public byte[] GetRandomKey() {
            if (fixedKey != null) return fixedKey;
            return GetRandomKey(16);
        }

        public byte[] GetRandomKey(int length) {
            if (fixedKey != null && fixedKey.Length == length) return fixedKey;
            return AESStream.GetRandomKey(length * 8);
        }

        public byte[] getRandomIV(int length) {
            if (fixedIV != null) return fixedIV;
            return AESStream.GetRandomIV();
        }

        public void SetFixedIV(byte[] iv) {
            this.fixedIV = iv;
        }

        public void SetFixedKey(byte[] key) {
            this.fixedKey = key;
        }

        public bool HasFixedKeyAndIV {
            get { return (fixedKey != null && fixedIV != null); }
        }
    }
}
